File: src/forager/ui/widgets/game_card.py
from __future__ import annotations
from PySide6.QtCore import Qt, Signal, QRectF, QTimer
from PySide6.QtGui import (
    QColor,
    QPen,
    QPainter,
    QPainterPath,
    QFont,
    QFontMetrics,
    QPixmap,
    QLinearGradient,
    QBrush,
    QConicalGradient,
)
from PySide6.QtWidgets import QWidget

from forager.core.game import Game
from forager.artwork import pipeline as art
from forager.ui.fonts import UI_FONT
from forager.ui.theme import C

_RADIUS = C.RADIUS

_HOVER_ZOOM = 0.045  # cover scales to 1.045 while hovered
_HOVER_DURATION = 0.18  # seconds for the fade in/out
_COMET_PERIOD = 1.0  # seconds per border orbit


def _alpha(color: QColor, a: int) -> QColor:
    return QColor(color.red(), color.green(), color.blue(), int(a))


def _ease_out_cubic(t: float) -> float:
    return 1.0 - (1.0 - t) ** 3


class GameCard(QWidget):
    clicked = Signal(object)
    activated = Signal(object)

    def __init__(
        self,
        game: Game,
        parent=None,
        card_w: int = CARD_W,
        card_h: int = CARD_H,
        fit_art: bool = False,
    ):
        super().__init__(parent)
        self.game = game
        self._focused = False
        self._art: QPixmap | None = None
        self._fit_art = fit_art

        self.setFixedSize(card_w, card_h)
        self.setCursor(Qt.CursorShape.PointingHandCursor)
        self.setAttribute(Qt.WidgetAttribute.WA_Hover, True)
        self.setAttribute(Qt.WidgetAttribute.WA_OpaquePaintEvent, True)

        self._hprog = 0.0  # linear hover progress 0..1 (animated)
        self._hover_clock = 0.0  # seconds while hovered (drives the comet orbit)
        self._hover_anim = QTimer(self)
        self._hover_anim.setInterval(16)
        self._hover_anim.timeout.connect(self._hover_tick)

    def set_art(self, pix: QPixmap | None):
        self._art = pix
        self.update()

    def _overlay_visible(self) -> bool:
        return self._focused or self.underMouse()

    def _hover_wanted(self) -> bool:
        return self._focused or self.underMouse()

    def _hover_level(self) -> float:
        return _ease_out_cubic(self._hprog)

    def paintEvent(self, event):
        p = QPainter(self)
        p.setRenderHint(QPainter.RenderHint.Antialiasing)
        p.setRenderHint(QPainter.RenderHint.SmoothPixmapTransform)

        w, h = self.width(), self.height()
        rect = QRectF(self.rect()).adjusted(0.5, 0.5, -0.5, -0.5)
        path = QPainterPath()
        path.addRoundedRect(rect, _RADIUS, _RADIUS)
        p.setClipPath(path)

        p.fillRect(rect, QColor(C.COLOR_3))

        t = self._hover_level()
        draw_rect = self._draw_rect(rect, t)

        if self._art is not None and not self._art.isNull():
            if self._fit_art:
                pix = art.scaled(
                    self._art, round(draw_rect.width()), round(draw_rect.height())
                )
                p.drawPixmap(
                    round(draw_rect.left() + (draw_rect.width() - pix.width()) / 2),
                    round(draw_rect.top() + (draw_rect.height() - pix.height()) / 2),
                    pix,
                )
            else:
                pix = art.scale_crop(
                    self._art, round(draw_rect.width()), round(draw_rect.height())
                )
                p.drawPixmap(draw_rect, pix, QRectF(pix.rect()))
        else:
            self._paint_placeholder(p, w, h, draw_rect)

        if self._overlay_visible():
            self._paint_overlay(p, w, h)

        p.setClipping(False)
        p.setBrush(Qt.BrushStyle.NoBrush)
        if t > 0:
            self._paint_hover_border(p, rect, t)

        if self._focused:
            p.setPen(QPen(QColor(C.ACCENT_1), 2))
            p.drawRoundedRect(rect, _RADIUS, _RADIUS)

    def _draw_rect(self, rect: QRectF, t: float) -> QRectF:
        z = 1.0 + _HOVER_ZOOM * t
        if abs(z - 1.0) < 1e-4:
            return rect
        nw, nh = rect.width() * z, rect.height() * z
        return QRectF(
            rect.center().x() - nw / 2.0,
            rect.center().y() - nh / 2.0,
            nw,
            nh,
        )

    def _paint_hover_border(self, p: QPainter, rect: QRectF, t: float):
        base = QColor(C.ACCENT_1)

        pen = QPen(_alpha(base, int(55 * t)))
        pen.setWidth(2)
        p.setPen(pen)
        p.drawRoundedRect(rect, _RADIUS, _RADIUS)

        angle = (self._hover_clock % _COMET_PERIOD) / _COMET_PERIOD * 360.0
        cg = QConicalGradient(rect.center(), angle)
        cg.setColorAt(0.0, _alpha(base, int(235 * t)))
        cg.setColorAt(0.14, _alpha(base, 0))
        cg.setColorAt(1.0, _alpha(base, 0))
        pen = QPen(QBrush(cg), 3.0)
        p.setPen(pen)
        p.drawRoundedRect(rect, _RADIUS, _RADIUS)

    def _fallback_art(self) -> QPixmap | None:
        size = (self.width(), self.height())
        if getattr(self, "_fallback", None) is None or self._fallback_size != size:
            self._fallback = art.placeholder_grid(self.game, size[0], size[1])
            self._fallback_size = size
        return self._fallback

    def _paint_placeholder(self, p: QPainter, w: int, h: int, target: QRectF):
        pix = self._fallback_art()
        if pix is not None and not pix.isNull():
            p.drawPixmap(target, pix, QRectF(pix.rect()))
            return
        self._paint_simple_placeholder(p, w, h)

    def _paint_simple_placeholder(self, p: QPainter, w: int, h: int):
        icon = art.load_icon(self.game, allow_network=False)
        if icon is not None:
            side = max(36, w // 3)
            icon = icon.scaled(
                side, side,
                Qt.AspectRatioMode.KeepAspectRatio,
                Qt.TransformationMode.SmoothTransformation,
            )
            p.drawPixmap(
                (w - icon.width()) // 2,
                max(6, h // 2 - icon.height() // 2 - 24),
                icon,
            )

        label = self.game.name.replace("/", " / ")
        font = QFont(UI_FONT, max(9, w // 21), QFont.Weight.Medium)
        p.setFont(font)
        fm = QFontMetrics(font)
        while fm.horizontalAdvance(label) > w - 24 and len(label) > 10:
            label = label[:-3] + "…"
        p.setPen(QColor(C.TEXT_DIM))
        p.drawText(
            QRectF(12, h - 40, w - 24, 24),
            Qt.AlignmentFlag.AlignCenter,
            label,
        )

    def _paint_overlay(self, p: QPainter, w: int, h: int):
        overlay_h = max(44, h // 5)
        grad = QLinearGradient(0, h - overlay_h, 0, h)
        grad.setColorAt(0, QColor(0, 0, 0, 0))
        grad.setColorAt(1, QColor(0, 0, 0, 205))
        p.fillRect(QRectF(0, h - overlay_h, w, overlay_h), grad)

        label = self.game.name.replace("/", " / ")
        font = QFont(UI_FONT, max(10, w // 19), QFont.Weight.DemiBold)
        p.setFont(font)
        fm = QFontMetrics(font)
        while fm.horizontalAdvance(label) > w - 20 and len(label) > 10:
            label = label[:-3] + "…"
        p.setPen(QColor(C.TEXT))
        p.drawText(
            QRectF(10, h - overlay_h + 8, w - 20, overlay_h - 12),
            Qt.AlignmentFlag.AlignLeft | Qt.AlignmentFlag.AlignVCenter,
            label,
        )

    def set_focused(self, focused: bool):
        self._focused = focused
        self._start_hover_anim()
        self.update()

    def enterEvent(self, event):
        super().enterEvent(event)
        self._start_hover_anim()
        self.update()

    def leaveEvent(self, event):
        super().leaveEvent(event)
        self._start_hover_anim()
        self.update()

    def _start_hover_anim(self):
        if not self._hover_anim.isActive():
            self._hover_anim.start()

    def _hover_tick(self):
        dt = self._hover_anim.interval() / 1000.0
        target = 1.0 if self._hover_wanted() else 0.0
        if self._hprog < target:
            self._hprog = min(target, self._hprog + dt / _HOVER_DURATION)
        elif self._hprog > target:
            self._hprog = max(target, self._hprog - dt / _HOVER_DURATION)
        if self._hover_wanted():
            self._hover_clock += dt
        else:
            self._hover_clock = 0.0
        self.update()
        if self._hprog == target and not self._hover_wanted():
            self._hover_anim.stop()

    def mousePressEvent(self, event):
        if event.button() == Qt.MouseButton.LeftButton:
            self.clicked.emit(self.game)

    def mouseDoubleClickEvent(self, event):
        if event.button() == Qt.MouseButton.LeftButton:
            self.activated.emit(self.game)

File: src/ui/widgets/GameCard.cpp
#include "ui/widgets/GameCard.h"
#include "ui/Theme.h"
#include "ui/Fonts.h"
#include <QPainter>
#include <QPainterPath>
#include <QFont>
#include <QFontMetrics>
#include <QConicalGradient>
#include <QLinearGradient>
#include <QMouseEvent>
#include <cmath>

using namespace theme::C;

static float easeOutCubic(float t) { return 1.0f - std::pow(1.0f - t, 3.0f); }
static float easeInOutCubic(float t) { return t < 0.5f ? 4*t*t*t : 1-std::pow(-2*t+2,3)/2; }

static QColor withAlpha(const QColor& c, int a) {
    return QColor(c.red(), c.green(), c.blue(), a);
}

GameCard::GameCard(const Game& game, int cardW, int cardH, QWidget* parent)
    : QWidget(parent), m_game(game), m_cardW(cardW), m_cardH(cardH)
{
    setFixedSize(cardW, cardH);
    setCursor(Qt::PointingHandCursor);
    setAttribute(Qt::WA_Hover, true);

    connect(&m_hoverAnim, &QTimer::timeout, this, &GameCard::hoverTick);
    m_hoverAnim.setInterval(16);
}

void GameCard::setArt(const QPixmap& pix) { m_art = pix; update(); }
void GameCard::setFocused(bool focused) { m_focused = focused; startHoverAnim(); update(); }

float GameCard::hoverLevel() const { return easeOutCubic(m_hprog); }

void GameCard::startHoverAnim() {
    if (!m_hoverAnim.isActive()) m_hoverAnim.start();
}

void GameCard::hoverTick() {
    float dt = m_hoverAnim.interval() / 1000.0f;
    bool wanted = m_focused || underMouse();
    float target = wanted ? 1.0f : 0.0f;

    if (m_hprog < target) m_hprog = std::min(target, m_hprog + dt / 0.18f);
    else if (m_hprog > target) m_hprog = std::max(target, m_hprog - dt / 0.18f);

    if (wanted) m_hoverClock += dt; else m_hoverClock = 0.0f;

    update();
    if (m_hprog == target && !wanted) m_hoverAnim.stop();
}

void GameCard::paintEvent(QPaintEvent*) {
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);
    p.setRenderHint(QPainter::SmoothPixmapTransform);

    int w = width(), h = height();
    QRectF rect(0.5, 0.5, w - 1, h - 1);
    QPainterPath path;
    path.addRoundedRect(rect, RADIUS, RADIUS);
    p.setClipPath(path);

    p.fillRect(rect, QColor(COLOR_3));

    float t = hoverLevel();
    float zoom = 1.0f + 0.045f * t;
    QRectF drawRect(rect.center().x() - rect.width()*zoom/2, rect.center().y() - rect.height()*zoom/2,
                    rect.width()*zoom, rect.height()*zoom);

    if (!m_art.isNull()) {
        QPixmap scaled = m_art.scaled(drawRect.size().toSize(), Qt::IgnoreAspectRatio, Qt::SmoothTransformation);
        int ox = (scaled.width() - (int)drawRect.width()) / 2;
        int oy = (scaled.height() - (int)drawRect.height()) / 2;
        p.drawPixmap(drawRect, scaled, QRectF(ox, oy, drawRect.width(), drawRect.height()));
    } else {
        paintPlaceholder(p, w, h, drawRect);
    }

    bool overlayVisible = m_focused || underMouse();
    if (overlayVisible) paintOverlay(p, w, h);

    p.setClipping(false);
    if (t > 0) paintHoverBorder(p, rect, t);
    if (m_focused) {
        p.setPen(QPen(QColor(ACCENT_1), 2));
        p.drawRoundedRect(rect, RADIUS, RADIUS);
    }
}

void GameCard::paintPlaceholder(QPainter& p, int w, int h, const QRectF& target) {
    Q_UNUSED(target);
    p.fillRect(QRectF(0, 0, w, h), QColor(COLOR_3));

    QFont font(fonts::UI_FONT, std::max(9, w / 21), QFont::Medium);
    p.setFont(font);
    QFontMetrics fm(font);
    QString label = m_game.name();
    while (fm.horizontalAdvance(label) > w - 24 && label.length() > 10)
        label = label.left(label.length() - 3) + QStringLiteral("\u2026");
    p.setPen(QColor(TEXT_DIM));
    p.drawText(QRectF(12, h - 40, w - 24, 24), Qt::AlignCenter, label);
}

void GameCard::paintOverlay(QPainter& p, int w, int h) {
    int overlayH = std::max(44, h / 5);
    QLinearGradient grad(0, h - overlayH, 0, h);
    grad.setColorAt(0, QColor(0, 0, 0, 0));
    grad.setColorAt(1, QColor(0, 0, 0, 205));
    p.fillRect(QRectF(0, h - overlayH, w, overlayH), grad);

    QFont font(fonts::UI_FONT, std::max(10, w / 19), QFont::DemiBold);
    p.setFont(font);
    QFontMetrics fm(font);
    QString label = m_game.name();
    while (fm.horizontalAdvance(label) > w - 20 && label.length() > 10)
        label = label.left(label.length() - 3) + QStringLiteral("\u2026");
    p.setPen(QColor(TEXT));
    p.drawText(QRectF(10, h - overlayH + 8, w - 20, overlayH - 12),
               Qt::AlignLeft | Qt::AlignVCenter, label);
}

void GameCard::paintHoverBorder(QPainter& p, const QRectF& rect, float t) {
    QPen pen(withAlpha(QColor(ACCENT_1), (int)(55 * t)));
    pen.setWidth(2);
    p.setPen(pen);
    p.setBrush(Qt::NoBrush);
    p.drawRoundedRect(rect, RADIUS, RADIUS);

    float angle = std::fmod(m_hoverClock, 1.0f) / 1.0f * 360.0f;
    QConicalGradient cg(rect.center(), angle);
    cg.setColorAt(0.0, withAlpha(QColor(ACCENT_1), (int)(235 * t)));
    cg.setColorAt(0.14, withAlpha(QColor(ACCENT_1), 0));
    cg.setColorAt(1.0, withAlpha(QColor(ACCENT_1), 0));
    p.setPen(QPen(QBrush(cg), 3.0f));
    p.drawRoundedRect(rect, RADIUS, RADIUS);
}

void GameCard::enterEvent(QEnterEvent*) { startHoverAnim(); update(); }
void GameCard::leaveEvent(QEvent*) { startHoverAnim(); update(); }

void GameCard::mousePressEvent(QMouseEvent* e) {
    if (e->button() == Qt::LeftButton) emit clicked(m_game);
}
void GameCard::mouseDoubleClickEvent(QMouseEvent* e) {
    if (e->button() == Qt::LeftButton) emit activated(m_game);
}
